lib/galaxy/tools/parameters/output_collect.py

```diff
@@ -425,7 +550,80 @@ def _compose(f, g):
     return lambda x: f(g(x))
 
 
-class CollectedDatasetMatch( object ):
+class JsonCollectedDatasetMatch( object ):
+
+    def __init__( self, as_dict, collector, filename, path=None ):
+        self.as_dict = as_dict
+        self.collector = collector
+        self.filename = filename
+        self.path = path
+
+    @property
+    def designation( self ):
+        # If collecting nested collection, grap identifier_0,
+        # identifier_1, etc... and join on : to build designation.
+        element_identifiers = self.raw_element_identifiers
+        if element_identifiers:
+            return ":".join(element_identifiers)
+        elif "designation" in self.as_dict:
+            return self.as_dict.get( "designation" )
+        elif "name" in self.as_dict:
+            return self.as_dict.get( "name" )
+        else:
+            return None
+
+    @property
+    def element_identifiers( self ):
+        return self.raw_element_identifiers or [self.designation]
+
+    @property
+    def raw_element_identifiers( self ):
+        identifiers = []
+        i = 0
+        while True:
+            key = "identifier_%d" % i
+            if key in self.as_dict:
+                identifiers.append(self.as_dict.get(key))
+            else:
+                break
+            i += 1
+
+        return identifiers
+
+    @property
+    def name( self ):
+        """ Return name or None if not defined by the discovery pattern.
+        """
+        name = None
+        if "name" in self.as_dict:
```

Comment:
Yeah - these were copied and pasted from the variant below that was reading regexes. I should have thought that through and made them a bit cleaner.